### PR TITLE
Adds possibility to read '.gz' files when using local firehose

### PR DESCRIPTION
**Pull request by se7entyse7en**

### CODE REVIEW

File: server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java [hunk 1]
```diff
@@ -45,93 +52,100 @@
...
+			    }
+
+			    @Override
+			    public LineIterator next()
+			    {
+				    final File f = files.poll();
+				    try
+				    {
+					    InputStream innerInputStream = new FileInputStream(f);
+					    final InputStream outerInputStream = f.getName().endsWith(".gz")
```
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java]: suggest `com.metamx.common.CompressionUtils#isGz`

Comment (drcrallen) [inline server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java]: Also proposing

```
InputStream inputStream = new FileInputStream(f);
if(CompressionUtils.isGz(f.getName()))
{
    inputStream = CompressionUtils.gzipInputStream(inputStream)
}
```

File: server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java [hunk 2]
```diff
@@ -45,93 +52,100 @@
...
+		              ? CompressionUtils.gzipInputStream(innerInputStream)
+		              : innerInputStream;
+
+					    return IOUtils.lineIterator(
```
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java]: To retain prior behavior, can this just be

```
return IOUtils.lineIterator(inputStream, Charsets.UTF_8);
```

File: server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java [hunk 3]
```diff
@@ -45,93 +52,100 @@
...
+		              new BufferedReader(
+		                  new InputStreamReader(outerInputStream, Charsets.UTF_8)));
+				    } catch (Exception e)
+				    {
+					    throw Throwables.propagate(e);
```
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java]: exception should close inputStream (safely)

**Comment by drcrallen** [comment]
Hi @se7entyse7en, thanks for the PR!

Can you add a unit test that confirms this works?

File: server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java (hunk 2 above)
Comment (se7entyse7en) [inline server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java]: Doesn't this lead to some performance issue? AFAIK wrapping a BufferedReader around Reader is advisable.

File: server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java [hunk 4]
```diff
@@ -45,93 +53,122 @@
...
+				    {
+					    log.warn(e, "Failed to read file [%s]", f.getName());
+					    if (rawInputStream != null)
+					    {
+						    try
+						    {
+							    rawInputStream.close();
+						    } catch (IOException ioe)
+						    {
+							    Throwables.propagate(ioe);
```
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java]: suggest `e.addSuppressed(ioe); throw Throwables.propagate(e);`

File: server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java [hunk 5]
```diff
@@ -45,93 +53,122 @@
...
+						    logMessage = "Reading file [%s]";
+						    inputStream = rawInputStream;
+					    }
+
+					    log.info(logMessage, f.getName());
+
+					    return IOUtils.lineIterator(
+		              new BufferedReader(
+		                  new InputStreamReader(inputStream, Charsets.UTF_8)));
+				    } catch (Exception e)
```
Comment (drcrallen) [inline server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java]: Do any of the checked exceptions include InterruptedException? If so it should be handled in a way that does not loose the interrupted status

**Comment by drcrallen** [comment]
@se7entyse7en it is possible that a certain change may have performance improvements in some areas. But to retain the prior behavior you should returning `IOUtils.lineIterator(inputStream, Charsets.UTF_8);`

Or submit proper benchmarking showing the impact on heap and performance.

**Comment by fjy** [comment]
:+1:

File: server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java [hunk 6]
```diff
@@ -45,93 +53,122 @@
...
-          @Override
-          public void remove()
-          {
-            throw new UnsupportedOperationException();
-          }
-        },
-        firehoseParser
-    );
-  }
+	private static final EmittingLogger log = new EmittingLogger(LocalFirehoseFactory.class);
```
Comment (himanshug) [inline server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java]: it appears you changed spaces to tabs. druid uses 2 spaces for indentation. pls use appropriate formatter as mentioned in https://github.com/druid-io/druid/blob/master/CONTRIBUTING.md

File: server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java [hunk 7]
```diff
@@ -45,93 +53,122 @@
...
+					    {
+						    logMessage = "Reading gzipped file [%s]";
+						    inputStream = CompressionUtils.gzipInputStream(rawInputStream);
+					    } else
+					    {
+						    logMessage = "Reading file [%s]";
+						    inputStream = rawInputStream;
+					    }
+
+					    log.info(logMessage, f.getName());
```
Comment (himanshug) [inline server/src/main/java/io/druid/segment/realtime/firehose/LocalFirehoseFactory.java]: can we just put the log.info in both places above instead?

**Comment by fjy** [comment]
@se7entyse7en can we finish this up?

**Comment by gianm** [comment]
Similar functionality was added in another PR (I don't have the number handy) and then improved in #5586 to handle more formats.

**Pull request closed without merging**